### PR TITLE
Zome macro

### PR DESCRIPTION
**Pull request by sphinxc0re**

### CODE REVIEW

File: tests/integration_test.rs [hunk 1]
```diff
@@ -1,33 +1,29 @@
...
+    // Call the exposed wasm function that calls the Commit API function
+    let result = hc.call("test_zome", "test_cap", "check_global", r#"{}"#);
+    assert!(result.unwrap().is_empty());
 
-  let test_logger = test_logger.lock().unwrap();
+    let test_logger = test_logger.lock().unwrap();
 
-  println!("{:?}", *test_logger);
+    println!("{:?}", *test_logger);
 }
```
Comment (lucksus) [inline tests/integration_test.rs]: I'd like to see the zome function `send_tweet` created by the macro tested here...

File: src/init_globals.rs [hunk 2]
```diff
@@ -2,33 +2,30 @@
 
 use holochain_wasm_utils::try_deserialize_allocation;
 
-extern {
-  fn hc_init_globals(encoded_allocation_of_input: u32) -> u32;
+extern "C" {
```
Comment (Connoropolous) [inline src/init_globals.rs]: Just a question, what does the "C" next to `extern` or `pub` do?

**Comment by Connoropolous** [comment]
Has there been any discussion around what to do for an HDK in another language besides Rust, without macros?

**Comment by Connoropolous** [comment]
Actually, I just had an idea...  command line utilities could be written, that then get called on in the .build files... which do what this macro does for Rust.

**Comment by lucksus** [comment]
No discussion about that yet, @Connoropolous. I would encourage these discussions and also encourage starting a TypeScript HDK. Just saw this today: https://github.com/holochain/holochain-proto-types/blob/master/index.d.ts. Seems like @willemolding and @maackle have started..? :)

**Comment by Connoropolous** [comment]
That may prove very useful... I think is in its intention meant to match holochain-proto though. 

Also we should be careful about calling it typescript... it's technically AssemblyScript, which I was reading more about today on their wiki: https://github.com/AssemblyScript/assemblyscript/wiki

It states: "AssemblyScript compiles a strictly-typed subset of TypeScript (a typed superset of JavaScript) to WebAssembly ahead of time (see: Limitations). It's not really a language on its own, though it provides several new WebAssembly-specific types and built-ins, "
lol, so a subset of a superset

But yeah, I may open up an hdk-assemblyscript repo

**Comment by maackle** [comment]
Yeah holochain-proto-types is just the TypeScript definition files for the holochain API, so you get compiler errors if you botch things at the type level. We hope they can be used directly with AssembyScript but we don't know how it really works.

File: wasm-test/src/lib.rs [hunk 3]
```diff
@@ -1,21 +1,41 @@
...
-  return 0;
+    return 0;
+}
+
+#[derive(Serialize)]
+struct TweetResponse {
+    ok: bool,
+}
+
+zome_functions! {
```
Comment (Connoropolous) [inline wasm-test/src/lib.rs]: So all Zome functions will have to be defined like this?

File: src/macros.rs [hunk 4]
```diff
@@ -0,0 +1,34 @@
+#[macro_export]
```
Comment (Connoropolous) [inline src/macros.rs]: Some explanatory comments plz!

File: wasm-test/src/lib.rs (hunk 3 above)
Comment (sphinxc0re) [reply wasm-test/src/lib.rs]: Yes. All of the top level zome functions will be written inside the macro

File: src/macros.rs (hunk 4 above)
Comment (sphinxc0re) [reply src/macros.rs]: Can do. What kind of explanation do you want me to write?

Comment (Connoropolous) [reply src/macros.rs]: A combination of: 
1. what each line does
2. a very quick explainer of the thing as a whole... how it functions and what it enables

Comment (Connoropolous) [reply src/macros.rs]: Especially the ParamStruct + Serialize/Deserialize bits

Comment (sphinxc0re) [reply src/macros.rs]: Okay, I'll write a doc comment about the usage

**Comment by lucksus** [comment]
Would love to merge this. Is the doc @sphinxc0re pushed enough for you @Connoropolous?

**Comment by Connoropolous** [comment]
I was hoping for line comments too, but yeah, it's good. I can add those at a later time too

**Comment by ddd-mtl** [comment]
So are we counting Connor as 2nd approver now? I'm not too happy about this being merged in.

**Comment by Connoropolous** [comment]
It might be helpful when you say "I'm not too happy" to clarify what it is
precisely that you're not too happy about

**Comment by ddd-mtl** [comment]
Yes, I am planning that. I wanted to share this first, since I am not sure when I will find the time to do that. Next time I'll add that I'm planning to write a follow up explaining why.

**Comment by ddd-mtl** [comment]
Here is a list of things I do not like:
 1. PR mixed in with rustfmt reformatting. Hard to see what actually changed.
 2. Unilateral reformatting without team approval. This is partly a consequence of unilateral decision to create this repo in the first place.
 3. Bloating pre-existing tests instead of making new ones.
4. No proper test for the `zome_functions!` macro. That tweet test function doesn't test serializing/deserializing from wasm memory which is the whole point of the macro. Finding out that the `zome_functions!` was indeed buggy when looking in the commit_entry() PR proves my point.
5. wasm lookup path changed from `debug` to `release` for unknown reasons.
6. No explanation in the PR description to help understand the previous points.

**Comment by Connoropolous** [comment]
In that case, I very much agree. 

I actually mentioned in a message to which I didn't get a reply, something here: https://chat.holochain.net/holo/pl/5fzh5m8q17b3x8efx3d755e9xa was something like what I mentioned the issue?

**Pull request merged**